Implemented additional caching features for #530

byko3y opened the pull request:
Parsing and application of no-cache and private directives with arguments.
Additional validation of Cache-Control header.
New cache_control_ignore, cache_resp_hdr_del configuration options.
Fixed corruption of cached responses with empty header value.

Fixed all the comments from the https://github.com/tempesta-tech/tempesta/pull/1539 here.

byko3y commented:
The original version of the pull request is not compilable because of excessive curly brackets. Somehow git brought my working copy 1 commit before the final version so I verified the final version itself. Good news is the last commit had only cosmetic changes so it's not a big deal.

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -2238,13 +2321,22 @@ __req_parse_authorization(TfwHttpReq *req, unsigned char *data, size_t len)
+	}
 	__FSM_STATE(Req_I_CC) {
-		WARN_ON_ONCE(parser->_acc);
+		WARN_ON_ONCE(parser->cache_control.dir_flag);
```
This is just not addressed comment https://github.com/tempesta-tech/tempesta/pull/1539/files#r775029875 . Please review the review notes once more. In short: you shouldn't fit new logic into existing states, most likely you need several new states.Introducing new states hurts, but not so much as introducing a lot of complexity with many if's and extra variables.

krizhanovsky replied:
This is just not addressed comment https://github.com/tempesta-tech/tempesta/pull/1539/files#r775029875 . Please review the review notes once more. In short: you shouldn't fit new logic into existing states, most likely you need several new states.Introducing new states hurts, but not so much as introducing a lot of complexity with many if's and extra variables.

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -2265,45 +2359,59 @@ __req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len)
 
 	__FSM_STATE(Req_I_CC_Flag) {
-		WARN_ON_ONCE(!parser->_acc);
+		/* A start of a common flag successfully detected */
```
What is a "common flag"? The comment is very confusing in context of the state name `Req_I_CC_Flag`, where "flag" is about the parser flag, see https://github.com/tempesta-tech/tempesta/pull/1539/files#r774807540

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -2312,65 +2420,102 @@ __req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len)
-		return CSTR_NEQ;
+		if (IS_CRLF(c)) {
+			if (!parser->cache_control.filled)
+				__FSM_EXIT(TFW_BLOCK);
```
A comment is required - why the success of the whole Cache-Control depends on cache_control.filled ? Actually, I believe the variable is just a variant of push down just to avoid new states, which makes the code overcomplicated and probably slower because we have extremely fast FSM technology

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8587,59 +8792,133 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+		if (IS_WS(c) || c == ',')
+			__FSM_I_MOVE_fixup(Resp_I_CC_Flag_Argument_Token, 1, 0);
+		if (c == '"')
+			__FSM_I_MOVE_fixup(Resp_I_EoT, 1, 0);
```
We should move `"` parsing into a separate state

krizhanovsky replied:
It's still worth to move `"` parsing to a separate small state

byko3y replied:
Sorry, I don't understand what part of the algorithm should be separated into the new state. To me it already looks like the most simple character probing possible.

krizhanovsky replied:
Yes, the state is simple, and good enough. I was confused with that it allows e.g. "Cache-Control: public, no-cache="Set-Cookie,"` (`"` just after `,`), but I believe that's OK.

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8587,59 +8792,133 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+			 &resp->private_tokens : &resp->no_cache_tokens;
+		if (__str_array_add_token(msg, tokens, p, __fsm_sz, false))
+			__FSM_EXIT(TFW_BLOCK);
+		/* Header name comparision requires a trailing semicolumn */
```
```suggestion
		/* Header name comparison requires a trailing colon */
```

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8587,59 +8792,133 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+			__FSM_EXIT(TFW_BLOCK);
+		/* Header name comparision requires a trailing semicolumn */
+		if (__str_array_add_token(msg, tokens, ":", 1, true))
+			__FSM_EXIT(TFW_BLOCK);
```
Do not make a separate call with the generic allocation and TfwStr logic. You know precisely on the `__str_array_add_token()` call above that you finished to parse a header name and you can call the function with a flag, which will add the required `:`

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8587,59 +8792,133 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
 		__FSM_I_JMP(Resp_I_Ext);
 	}
 
+	__FSM_STATE(Resp_I_Flag_Maybe_Qal) {
```
Please make a comment for the state referencing the RFC chapters requiring header names parsing on `no-cache` and `private`

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8680,56 +8964,77 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+		 * Here we just skip all the tokens, double quotes and
+		 * equality signs.
+		 */
 		/* TODO: process cache extensions. */
```
Please remove the comment as well

krizhanovsky reviewed (changes requested):
There are still several bugs and more work is required to for the FSM workflow simplification (basically already commented in https://github.com/tempesta-tech/tempesta/pull/1539). The FSM issues are somewhat linked with https://github.com/tempesta-tech/tempesta/pull/1538 , so please talk with @Dmitry-Gouriev - we already had conversations about automata constructions.

Also #530 requires several functional tests tests including https://github.com/tempesta-tech/tempesta/issues/810#issuecomment-966690711 (note the `cache_bypass` and `cache_fullfil` configuration options, which can not be tested with unit tests) and Wiki update.

Dmitry-Gouriev commented on fw/http_parser.c:
```diff
@@ -767,39 +773,116 @@ __hbh_parser_add_data(TfwHttpMsg *hm, char *data, unsigned long len, bool last)
+
+/* similar to tfw_http_msg_grow_hdr_tbl */
+int
+__str_array_grow(TfwHttpMsg *hm, TfwStrArray *array)
```
Stange rename hdr -> hbh_hdr.

Dmitry-Gouriev commented on fw/http_parser.c:
```diff
@@ -2265,45 +2359,59 @@ __req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len)
 
 	__FSM_STATE(Req_I_CC_o) {
+		if (unlikely(!parser->cache_control.dir_allowed))
+			__FSM_EXIT(TFW_BLOCK);
```
Hello! I wonder a little that the conditions `!parser->cache_control.dir_allowed` is tested at several states.
It is set to `true` at `Req_I_CC_start`. It can become `false` during parsing? Or the parser can enter these states avoiding `Req_I_CC_start` and with `false` value for condition? Believing to the given name `dir_allowed`, the check probably better be outside this function at all?

Dmitry-Gouriev commented on fw/http_parser.c:
```diff
@@ -2312,65 +2420,102 @@ __req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len)
+		if (parser->cache_control.dir_flag != 0) {
+			/* Req_I_EoT will update the field at comma */
+			parser->cache_control.dir_allowed = false;
+		}
```
Hello! It seems I do not understand a conception of `dir_allowed` flag (or can hardly agree with it).

The flag is set to `true` in two states `Req_I_CC_start` and `Resp_I_EoT`, reset to `false` in several other states and is checked in many other states.

It seems that the flag is used to disable (make disallowed) a subset of states.

Is not it better to provide the same by hardcoding transitions between states with `__FSM_JPM`, `__FSM_I_MOVE_xxxx`?

We can control state transition hardly and `__FSM_START` is to restore that state on the next call of the function rather than to enter a random unpredictable state.

Dmitry-Gouriev replied:
Looking from more general position, I'd use states to encode parser positions inside text being parsed and flags and variables to store information which was detected in the past and needed in the future. However performance consideration could dictate something different.

krizhanovsky commented on fw/hpack.c:
```diff
@@ -1888,6 +1888,11 @@ do {									\
+			/* We are going to skip the HPACK_STATE_VALUE_TEXT,
+			   but still need the newline for a correct HTTP/1 */
+			EXPAND_DATA(S_CRLF, SLEN(S_CRLF));
+		}
```
This is for retrieving headers with empty values from the cache, right? In other words I believe this is the part of https://github.com/tempesta-tech/tempesta/pull/1551 .

Please move the fix to #1551 AND write a better comment:
```
/*
 * The header has no value, so we are going to skip the HPACK_STATE_VALUE_TEXT,
 * but still need to write CRLF at the end of the header line for a correct HTTP/1
 */
```

krizhanovsky commented on fw/hpack.c:
```diff
@@ -1888,7 +1888,16 @@ do {									\
 		NEXT_STATE(HPACK_STATE_VALUE_TEXT);
 
-		GET_NEXT_DATA(src >= last);
+		if (src >= last) {
```
It still makes sense to leave the condition with `unlikely`

krizhanovsky commented on fw/vhost.h:
```diff
@@ -117,6 +141,8 @@ typedef struct {
 	size_t			capo_sz;
 	size_t			nipdef_sz;
 	TfwCaPolicy		**capo;
+	TfwCaPolicyTokens	capo_hdr_del;
```
the comment is still not addressed

byko3y replied:
It is addressed: TfwCaPolicyTokens no longer exists, instead there is "TfwCaToken *capo_hdr_del".

krizhanovsky replied:
I meant that why not to combine all caching policy stuff, including the new configuration directives and `cc_ignore`, into TfwCaPolicy. In this sense there is no difference between TfwCaPolicyTokens and TfwCaToken.

However, I see that `capo` and `capo_hdr_del` are arrays of different sizes, not single structures, which can be combined. So we're good with the current design.

Dmitry-Gouriev commented on fw/http_parser.h:
```diff
@@ -127,7 +127,6 @@ typedef struct {
 			char		filled;
-			char		dir_allowed;
 		} cache_control;
 	};
```
I finally do not fully understand an idea of the `dir_flag `.
It seems to be unneccessary in request parser (you can make additions (`|=`) directly to `req->cache_ctl.flags` instead of `dir_flag `.
It used in 2 conditions in reqponse parser:
```C
Line 8750: 			if (parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE)
Line 8787: 		tokens = parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE ?
```
This is because the states `Resp_I_Flag_Maybe_Qal` and `Resp_I_CC_Flag_Argument_Token` are shared between several Cache-Control directives. You can divide these states into several ones to aviod sharing. If you do not like new states... Anyway, it seems, it would be simplier to use the `dir_flag` **only** to cache the directive flag for these checks, not to set `resp->cache_ctl.flags`.

byko3y replied:
Actually there is another condition in the Req_I_CC_Flag/Resp_I_Flag, so we cannot set the req->cache_ctl.flags immediately. However, I agree it's not needed anywhere else and Resp_I_Flag_Maybe_Qal should be segregated into two states.

Dmitry-Gouriev replied:
@byko3y 
If you mean
```
WARN_ON_ONCE(!parser->cache_control.dir_flag);
```
this is about internal check for algorithm consistency rather than to parse the text.
However this consistence condition could be proved staticaly.
Is there any way to reach these states without `reg_flag` set?

byko3y replied:
Everything can be proven statically, it's more about how complex the proof is. In this case I do agree the proof is quite simple so there's not much use for this particular statement.

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -2312,65 +2330,107 @@ __req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len)
+		if (IS_CRLF(c)) {
+			/* Cache-Control header cannot be empty */
+			if (!parser->cache_control.filled)
+				__FSM_EXIT(TFW_BLOCK);
```
+1 for @Dmitry-Gouriev 's comments. `dir_flag` is just a replacement for old `_acc` and I'm good with it. But Dmitry pointed out that probably we can directly set the flags, w/o collecting them in `_acc`. `_acc` stands for "accumulator" and the primary target for the field is to accumulate multi-digit number values, which seem not necessary for flags.

`cache_control.filled` still looks unnatural, but there isn't so many code around it and now FSM looks much more consistent (so I closed #1539). However, probably we can determine whether Cache-Control had a directive with introducing only one new state. In this case we won't need `filled`.

If we can get rid of both the `filled` and `dir_flag`, then we can get rid of `Req_I_CC_start` and, with one new state instead of `filled`, we'll have the same number of states and more straightforward and likely faster FSM.

BTW since `Req_I_CC_start` is executed only once on very first FSM entrance, we can initialize the variable just at the beginning of the function and we don't need a state for it.

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -6013,24 +6089,49 @@ __h2_req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len,
 			__FSM_I_JMP(Req_I_CC);
-		return CSTR_NEQ;
+		}
+		/* Should also check for parser->cache_control.filled at CRLF */
```
we have no CRLFs in h2

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8600,45 +8710,116 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+			__FSM_I_MOVE_fixup(Resp_I_CC_Opening_Quote, 1, 0);
+		}
+		/* not qualified, fall though */
+	}
```
an empty line is wished here

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8600,45 +8710,116 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+			if (parser->cache_control.dir_flag == TFW_HTTP_CC_PRIVATE)
+				resp->cache_ctl.flags |= TFW_HTTP_CC_PRIVATE_QUAL;
+			else
+				resp->cache_ctl.flags |= TFW_HTTP_CC_NO_CACHE_QUAL;
```
Ok, I see the usage of `dir_flag` versus direct update in `cache_ctl.flags`. Otherwise we'll need at least one more state them. However, neither of the *_QUAL flags are actually used besides the tests, so they should be just removed (and the tests reworked to test the things in a different way).

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8600,45 +8710,116 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+		if (c != '"')
+			__FSM_EXIT(TFW_BLOCK);
+		else
+			__FSM_I_MOVE_fixup(Resp_I_CC_Flag_Argument_Token, 1, 0);
```
Usually, we don't use else in such cases, just
```C
if (c != '"')
    __FSM_EXIT(TFW_BLOCK);
__FSM_I_MOVE_fixup(Resp_I_CC_Flag_Argument_Token, 1, 0);
```

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8600,45 +8710,116 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+					       ":", 1))
+		{
+			__FSM_EXIT(TFW_BLOCK);
+		}
```
Looks much better now: shorter and more efficient!

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8680,56 +8864,83 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
 		if (IS_WS(c))
 			__FSM_I_MOVE_fixup(Resp_I_EoT, 1, TFW_STR_OWS);
 
-		parser->_acc = 0; /* reinit for next token */
```
Don't you need to reinit `dir_flag` as well?

krizhanovsky reviewed (changes requested):
The FSM is much better now and we're almost "there".

Functional tests and wiki are still in TODO. It'd better to make the tests and make sure that they do pass before the final review and merge of the code.

byko3y commented on fw/http_parser.c:
```diff
@@ -2471,39 +2378,47 @@ __req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len)
+
 		if (IS_TOKEN(c)) {
 			/* reinit for next token */
 			parser->cache_control.dir_flag = 0;
```
Agreed, RFC seems to forbid empty tokens, I'll fix that.

byko3y commented on fw/cache.c:
```diff
@@ -1679,6 +1724,38 @@ __cache_entry_size(TfwHttpResp *resp)
+		/* remove headers mentioned in cache_resp_hdr_del */
+		if (hdr_del_tokens.tokens) {
+			if (check_cfg_ignored_header(hdr, hdr_del_tokens.tokens,
+						     hdr_del_tokens.sz))
```
The nesting problem mostly comes from large tab width and small line width limit. However, in this particular case the nested block is small and fits well into 80 character limit.

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -5860,8 +5913,22 @@ __h2_req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len,
+
+	__FSM_STATE(Req_I_CC_start) {
+		if (IS_WS(c))
+			__FSM_I_MOVE(Req_I_CC_start);
```
yeah, it seems we had the issue in the parsing function because it didn't process WS at the beginning. However, it seems we can move the check and transition to `Req_I_CC`

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8600,45 +8699,140 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+		if (c == '=') {
+			/* Qualified form cancels out the general flag */
+			/*
+			 * Resp_I_CC_Opening_Quote => Resp_I_CC_Dir_Arg_Token =>
```
```suggestion
			 * Resp_I_CC_Dir_Opening_Quote => Resp_I_CC_Dir_Arg_Token =>
```

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8600,45 +8699,140 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+			__FSM_I_MOVE_fixup(Resp_I_CC_Dir_Opening_Quote, 1, 0);
+
+		__FSM_I_JMP(Resp_I_Flag);
+	}
```
This state is just the same as `Resp_I_Flag_No_Cache`, so the states should be unified

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8600,45 +8699,140 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+		if (c == ',')
+			__FSM_EXIT(TFW_BLOCK);
+		if (c == '"')
+			__FSM_I_MOVE_fixup(Resp_I_EoT, 1, 0);
```
This state is to allow `no-cache="Set-Cookie, "` - is this intentional? If yes, then comment is required. Otherwise, we can just omit the check and block `"` on the default transition to `TFW_BLOCK`

krizhanovsky reviewed (changes requested):
There are still things to fix and clanup

byko3y commented on fw/http_parser.c:
```diff
@@ -8566,13 +8646,32 @@ STACK_FRAME_NON_STANDARD(__resp_parse_age);
+			__FSM_I_MOVE_fixup(Req_I_CC_start, 1, TFW_STR_OWS);
+		/* Two consecutive commas not allowed */
+		if (c == ',')
+			__FSM_EXIT(TFW_BLOCK);
```
I did some brief benchmarks on this - after every unnecessary branch removed the `__resp_parse_cache_control` function became 0.8% faster. During `TEST_RUN(http_parser, fills_hdr_tbl_for_resp)` the `__resp_parse_cache_control` takes 0.1% of the run time (thus parser is 0.008% faster with modifications), and `TEST_RUN(http_parser, cache_control)`, which feeds nothing but "cache-control" headers to the parser, leads to `__resp_parse_cache_control` working 25% of time (parser 0.2% faster, probably same increase for `__req_parse_cache_control`).
I just don't feel like performance is an issue here. Readability and robustness are though.
Also, trailing pure conditions with same statements in branches are optimized into a single unconditional statement even at "-Og" optimization level.

krizhanovsky replied:
First of all, there are many questions about performance: how many times you ran the test? What the test is? Did you pick the *best* times for each of the tests? Did you test different hardware?

There is still no reason to have more code if you can have less code. More code tends to be harder to read and understand, especially if the code does senseless checks.

byko3y commented:
- Removed unnecessary conditions and reworked leading comma handling in cache-control parsing functions;
- Additional tests for commas in cache-control directives and their arguments;
- Reworked tfw_str_array_append_chunk();
- Reworked __tfw_strcmpspn() to match strings even if one of them lacks a trailing separator;
- Additional tests for both tfw_str_array_append_chunk and __tfw_strcmpspn().

Will update the copyright years once the final list of files is clear.

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -6014,23 +6080,37 @@ __h2_req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len,
+		if (IS_WS(c))
+			__FSM_H2_I_MOVE(Req_I_After_Comma);
+		if (c == ',')
+			__FSM_EXIT(TFW_BLOCK);
```
The comment above isn't addressed. Why do we need the check and why it's not enough just to check for token at the below?

krizhanovsky commented on fw/str.c:
```diff
@@ -670,6 +688,181 @@ tfw_str_add_duplicate(TfwPool *pool, TfwStr *str)
+ *                        len = 3,
+ *                        data = "tok",
+ *                        flags = 0 }     <= last uncompleted item
+ *          nchunks = 3
```
Please update the comment after the simplification. When the function has signature `tfw_str_array_append_chunk(TfwPool *pool, TfwStr *array, char *data, unsigned long len)` that won't be obvious how it's different from `tfw_str_add_compound()`/`__str_grow_tree()`, so please address this question (why do we need the new function) in the comment.

krizhanovsky reviewed (changes requested):
Unfortunately at least one more change and one more review is required. The main issue is still overcomplecated `tfw_str_array_append_chunk()`, which must be reworked. Several small cleanups are also required.

Please make the changes in a new commit, just like you did this time.

byko3y commented:
Hopefully resolved all the comments.

krizhanovsky commented on fw/str.c:
```diff
@@ -670,6 +688,139 @@ tfw_str_add_duplicate(TfwPool *pool, TfwStr *str)
+		shall_complete = false;
+	}
+
+	if (!len) {
```
```suggestion
	if (unlikely(!len)) {
```

krizhanovsky commented on fw/str.c:
```diff
@@ -670,6 +688,139 @@ tfw_str_add_duplicate(TfwPool *pool, TfwStr *str)
+	last_chunk->len = len;
+	last_chunk->data = data;
+	/*
+	 * We don't care about array->len, but the last_token->len is required
```
I think it makes sense to replace
```
We don't care about array->len, but
```
with
```
array->len has no sense since this is an array of independent strings. However,
```

krizhanovsky commented on fw/str.c:
```diff
@@ -670,6 +688,139 @@ tfw_str_add_duplicate(TfwPool *pool, TfwStr *str)
+	if (!last_chunk)
+		return -ENOMEM;
+
+	added_len += len;
```
`added_len` is always zero here, so just use `len` at the below and remove `added_len` completely

krizhanovsky commented on fw/http_parser.c:
```diff
@@ -8600,45 +8715,134 @@ __resp_parse_cache_control(TfwHttpResp *resp, unsigned char *data, size_t len)
+
+	__FSM_STATE(Resp_I_CC_Dir_Arg_Token) {
+		/* comma-separated field list (field-name/token) with optional
+		 * linear white space
```
Please start the sentence form Captital letter and end it with dot.

Also please add following comment here: "We do not set TfwStr->skb for tokens, since the tokens are used in read-only fashion to compare strings and are never rewriten (TfwStr->skb is used for rewriting raw data only)."

krizhanovsky reviewed (approved):
Good to merge after the small cleanups
